### PR TITLE
Basic uniform refinement algorithm

### PR DESCRIPTION
**Pull request by termi-official**
Seems to work fine but fills the ram really quickly due to massive duplication in the computations. Not sure how to address this in a cache-friendly way.

## TODO

* [x] Clarify docs further

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -155,13 +161,14 @@ function crinkle_clip(plotter::MakiePlotter{3,DH,T}, decision_fun) where {DH,T}
          plotter.dh,
          plotter.u,
          plotter.topology,
-         plotter.visible,
+         copy(plotter.visible),
```
Comment (koehlerson) [inline src/utils.jl:164]: why the copy?

File: src/utils.jl [hunk 2]
```diff
@@ -370,42 +377,44 @@ function _transfer_solution!(data,pv,fh,ip_geo,ip_field,cellset_,val_buffer,val,
...
     local_dof_range = Ferrite.dof_range(fh, field_name)
-    _processreturn = length(process(val_buffer))
-
-    current_vertex_index = 1
+    _processreturndim = length(process(val_buffer))
 
     cell_geo_ref = Ferrite.getcells(grid, cellset_[1])
 
-    Ferrite.reinit!(pv, Ferrite.getcoordinates(grid,cellset_[1]), Tensors.Vec{dim}(ref_coords[current_vertex_index,:]))
+    Ferrite.reinit!(pv, Ferrite.getcoordinates(grid,cellset_[1]), Tensors.Vec{dim}(ref_coords[1,:]))
```
Comment (koehlerson) [inline src/utils.jl:384]: if the first index is constant, can we resort to a vector instead of a matrix?

File: src/utils.jl [hunk 3]
```diff
@@ -370,42 +377,44 @@ function _transfer_solution!(data,pv,fh,ip_geo,ip_field,cellset_,val_buffer,val,
...
-                val_buffer[d] = Ferrite.function_value(pv, 1, @views(u[_celldofs_field[d,:]]))
-            end
-            val = process(val_buffer)
-            for d in 1:_processreturn
-                data[current_vertex_index,d] = val[d]
+
+        # Loop over the triangles of the cell and interpolate at the vertices
+        # TODO remove redundant function value calls
+        for triangle_index in triangles_on_cell(plotter, cell_idx)
+            for current_vertex_index in plotter.all_triangles[triangle_index]
```
Comment (koehlerson) [inline src/utils.jl:407]: `all_triangles` renders also the ones that aren't visible right? So isn't here the skip missing if its visible or not?

**Review by koehlerson — COMMENTED** [review]
few questions

File: src/utils.jl (hunk 1 above)
Comment (termi-official) [reply src/utils.jl:164]: I can make a separate PR for this bugfix if you want. The issue is that if you apply a crinkle clip and plot with the original plotter, then the visibility vector is messed up, leading to a different visualization result - but the method without "!" postfix should not modify the original object, right?

File: src/utils.jl (hunk 2 above)
Comment (termi-official) [reply src/utils.jl:384]: I think I do not understand the question. You mean that we should repalce the matrix with a vector of vectors?

Comment (koehlerson) [reply src/utils.jl:384]: ah nvm its just the initialization

File: src/utils.jl (hunk 3 above)
Comment (termi-official) [reply src/utils.jl:407]: Huh? I have not changed the logic here and thought that `vis_triangles` is the makie buffer for rendering where invisible triangles get the vertices 1,1,1?

Comment (koehlerson) [reply src/utils.jl:407]: ah right its above, nvm

File: src/utils.jl (hunk 1 above)
Comment (koehlerson) [reply src/utils.jl:164]: Ah true, I guess then its fine

**Comment by termi-official** [comment]
```julia
using FerriteViz;
include("docs/src/ferrite-examples/heat-equation.jl");
dh, u = manufactured_heat_problem(Hexahedron, Lagrange{3, RefCube, 2}(), 2);
plotter = FerriteViz.MakiePlotter(dh,u);
FerriteViz.transfer_solution(plotter, u);
@time FerriteViz.transfer_solution(plotter, u);
```

PR: 0.142724 seconds (41 allocations: 4.601 MiB)
Master: 0.138495 seconds (41 allocations: 4.601 MiB)

**Pull request merged**